File: runtime/go/prompty/model/agent_loop_engine.go
package prompty

// Provider-agnostic agent loop -- the canonical TurnConformance.run engine.
//
// This file owns the observable agent-loop contract that the cross-runtime
// @vector suite (schema/model/conformance/vectors/agent.tsp, stage "agent")
// asserts. It is deliberately provider-agnostic: the loop is driven by two
// abstract callbacks -- InvokeModel(conversation) -> AgentModelResponse (one LLM
// call) and DispatchTool(call) -> string (one tool execution) -- so the same
// engine backs every provider. Providers supply only the wire translation that
// turns their raw response into an AgentModelResponse; they never re-implement
// the loop, its accounting, or its event vocabulary. This mirrors the Python
// reference in prompty/core/agent_loop.py so every runtime shares one contract.
//
// Observable contract (verified against all 28 run vectors):
//   - Iterations counts LLM calls (not tool rounds).
//   - TotalMessages == len(conversation) + (1 if any tool round ran else 0).
//   - messages_updated.message_count == len(conversation) + 1 at emit time.
//   - Events are emitted in a fixed order: status -> tool_call_start ->
//     tool_result -> messages_updated -> optional steering status/
//     messages_updated -> done; cancelled replaces the tail on cancellation.
//   - Canonical message shapes: assistant-with-tool-calls
//     {role:"assistant", content:"", metadata:{tool_calls:[...]}}; tool result
//     {role:"tool", content:<str>, metadata:{tool_call_id:<id>}}; final answer
//     {role:"assistant", content:<str>}.
//
// Errors are returned as fields on AgentLoopResult (Error/ErrorType/ErrorReason)
// rather than raised, so the accumulated conversation and events remain
// observable on the failure path.

import (
	"encoding/json"
	"strings"
)

// AgentDefaultMaxIterations is the loop's default iteration ceiling.
const AgentDefaultMaxIterations = 10

// AgentSummaryPrefix marks a compaction summary system message.
const AgentSummaryPrefix = "[Summary of earlier conversation] "

// Canonical error markers. The vectors assert the class name for cancellation
// and guardrail denials, so these mirror the exception types used by the other
// runtimes without coupling the loop to them.
const (
	agentCancelledError = "CancelledError"
	agentGuardrailError = "GuardrailError"
)

// AgentToolCall is a single tool invocation requested by the model.
type AgentToolCall struct {
	Id   string
	Name string
	// Arguments is the raw JSON string exactly as the model emitted it.
	Arguments string
}

// AgentModelResponse is a normalized single-turn model response.
//
// RawToolCalls carries the provider's exact tool-call array so the assistant
// message's metadata.tool_calls round-trips byte-for-byte; when nil the engine
// reconstructs it from AgentToolCall fields.
type AgentModelResponse struct {
	Content      interface{}
	ToolCalls    []AgentToolCall
	RawToolCalls []interface{}
}

// AgentGuardrailDecision is the outcome of a guardrail check.
type AgentGuardrailDecision struct {
	Allowed bool
	Reason  interface{}
}

// AgentSteeringMessage is a steering message scheduled for injection before a
// given iteration.
type AgentSteeringMessage struct {
	InjectBeforeIteration int
	Role                  string
	Text                  string
}

// AgentLoopOptions bundles the abstract callbacks and scripted flags that drive
// one agent-loop run.
type AgentLoopOptions struct {
	InvokeModel      func(conversation []map[string]interface{}) AgentModelResponse
	DispatchTool     func(call AgentToolCall) string
	IsToolRegistered func(name string) bool
	MaxIterations    int
	InputGuardrail   func(conversation []map[string]interface{}) AgentGuardrailDecision
	OutputGuardrail  func(response AgentModelResponse) AgentGuardrailDecision
	ToolGuardrail    func(name string, arguments map[string]interface{}) AgentGuardrailDecision
	Steering         []AgentSteeringMessage
	CancelAt         string
	ContextBudget    *int
	Summarize        func(dropped []map[string]interface{}) string
}

// AgentLoopResult is the observable result of an agent-loop run.
type AgentLoopResult struct {
	Result             interface{}
	Iterations         int
	Conversation       []map[string]interface{}
	Events             []map[string]interface{}
	ToolRounds         int
	ToolsExecuted      int
	ToolExecutionOrder []string
	DeniedTools        []string
	TrimmedMessages    []map[string]interface{}
	Error              interface{}
	ErrorType          interface{}
	ErrorReason        interface{}
}

// TotalMessages is the conversation length plus the conformance +1 when tools
// ran.
func (r AgentLoopResult) TotalMessages() int {
	total := len(r.Conversation)
	if r.ToolRounds > 0 {
		total++
	}
	return total
}

func agentCopyMessage(message map[string]interface{}) map[string]interface{} {
	out := make(map[string]interface{}, len(message))
	for key, value := range message {
		out[key] = value
	}
	return out
}

func agentAssistantToolCallsMessage(response AgentModelResponse) map[string]interface{} {
	var toolCalls []interface{}
	if response.RawToolCalls != nil {
		toolCalls = response.RawToolCalls
	} else {
		toolCalls = make([]interface{}, 0, len(response.ToolCalls))
		for _, call := range response.ToolCalls {
			toolCalls = append(toolCalls, map[string]interface{}{
				"id":   call.Id,
				"type": "function",
				"function": map[string]interface{}{
					"name":      call.Name,
					"arguments": call.Arguments,
				},
			})
		}
	}
	return map[string]interface{}{
		"role":     "assistant",
		"content":  "",
		"metadata": map[string]interface{}{"tool_calls": toolCalls},
	}
}

func agentToolMessage(callId string, content string) map[string]interface{} {
	return map[string]interface{}{
		"role":     "tool",
		"content":  content,
		"metadata": map[string]interface{}{"tool_call_id": callId},
	}
}

func agentCharCount(messages []map[string]interface{}) int {
	total := 0
	for _, message := range messages {
		if content, ok := message["content"].(string); ok {
			total += len(content)
		}
	}
	return total
}

func agentParseArgs(arguments string) map[string]interface{} {
	if arguments == "" {
		return map[string]interface{}{}
	}
	var parsed interface{}
	if err := json.Unmarshal([]byte(arguments), &parsed); err != nil {
		return map[string]interface{}{}
	}
	if asMap, ok := parsed.(map[string]interface{}); ok {
		return asMap
	}
	return map[string]interface{}{}
}

func agentDefaultSummary(droppedUsers []map[string]interface{}) string {
	topics := make([]string, 0, len(droppedUsers))
	for _, message := range droppedUsers {
		if content, ok := message["content"].(string); ok {
			trimmed := strings.TrimSpace(content)
			if trimmed != "" {
				topics = append(topics, trimmed)
			}
		}
	}
	return AgentSummaryPrefix + "User asked about " + strings.Join(topics, "; ")
}

func agentMaybeTrim(
	conversation []map[string]interface{},
	contextBudget *int,
	summarize func(dropped []map[string]interface{}) string,
) []map[string]interface{} {
	if contextBudget == nil || agentCharCount(conversation) <= *contextBudget {
		return nil
	}

	var systems []map[string]interface{}
	var users []map[string]interface{}
	for _, message := range conversation {
		switch message["role"] {
		case "system":
			systems = append(systems, agentCopyMessage(message))
		case "user":
			users = append(users, message)
		}
	}

	var droppedUsers []map[string]interface{}
	var lastUser map[string]interface{}
	if len(users) > 0 {
		droppedUsers = users[:len(users)-1]
		lastUser = users[len(users)-1]
	}

	var summaryText string
	if summarize != nil {
		summaryText = summarize(droppedUsers)
	} else {
		summaryText = agentDefaultSummary(droppedUsers)
	}

	trimmed := make([]map[string]interface{}, 0, len(systems)+2)
	trimmed = append(trimmed, systems...)
	trimmed = append(trimmed, map[string]interface{}{"role": "system", "content": summaryText})
	if lastUser != nil {
		trimmed = append(trimmed, map[string]interface{}{"role": "user", "content": lastUser["content"]})
	}
	return trimmed
}

// RunAgentLoop runs the canonical agent loop and returns its observable result.
//
// CancelAt accepts the scripted positions "before_iteration" (before iteration
// 1), "before_iteration_<n>" (before iteration n), and "after_tool_<i>" (after
// the i-th tool of a round). The loop is deterministic: given the same
// callbacks and flags it always produces the same events and accounting.
func RunAgentLoop(messages []map[string]interface{}, opts AgentLoopOptions) AgentLoopResult {
	maxIterations := opts.MaxIterations
	if maxIterations <= 0 {
		maxIterations = AgentDefaultMaxIterations
	}

	result := AgentLoopResult{
		ToolExecutionOrder: []string{},
		DeniedTools:        []string{},
	}

	conversation := make([]map[string]interface{}, 0, len(messages))
	for _, message := range messages {
		conversation = append(conversation, agentCopyMessage(message))
	}

	emit := func(eventType string, data map[string]interface{}) {
		result.Events = append(result.Events, map[string]interface{}{"type": eventType, "data": data})
	}

	emit("status", map[string]interface{}{"message": "Starting agent loop"})

	if trimmed := agentMaybeTrim(conversation, opts.ContextBudget, opts.Summarize); trimmed != nil {
		conversation = trimmed
		result.TrimmedMessages = make([]map[string]interface{}, 0, len(trimmed))
		for _, message := range trimmed {
			result.TrimmedMessages = append(result.TrimmedMessages, agentCopyMessage(message))
		}
	}

	steeringPending := make([]AgentSteeringMessage, len(opts.Steering))
	copy(steeringPending, opts.Steering)

	registered := opts.IsToolRegistered
	if registered == nil {
		registered = func(string) bool { return true }
	}

	for {
		iterationNumber := result.Iterations + 1

		if opts.CancelAt == "before_iteration" && iterationNumber == 1 {
			emit("cancelled", map[string]interface{}{"reason": "Cancellation requested before first iteration"})
			result.Error = agentCancelledError
			result.Conversation = conversation
			return result
		}
		if opts.CancelAt == agentBeforeIterationKey(iterationNumber) {
			emit("cancelled", map[string]interface{}{
				"reason": "Cancellation requested before iteration " + itoa(iterationNumber),
			})
			result.Error = agentCancelledError
			result.Conversation = conversation
			return result
		}

		var toInject []AgentSteeringMessage
		var remaining []AgentSteeringMessage
		for _, steer := range steeringPending {
			if steer.InjectBeforeIteration == iterationNumber {
				toInject = append(toInject, steer)
			} else {
				remaining = append(remaining, steer)
			}
		}
		if len(toInject) > 0 {
			steeringPending = remaining
			emit("status", map[string]interface{}{"message": "Injecting steering message"})
			for _, steer := range toInject {
				conversation = append(conversation, map[string]interface{}{"role": steer.Role, "content": steer.Text})
			}
			emit("messages_updated", map[string]interface{}{"message_count": len(conversation) + 1})
		}

		if opts.InputGuardrail != nil {
			decision := opts.InputGuardrail(conversation)
			if !decision.Allowed {
				result.Error = agentGuardrailError
				result.ErrorReason = decision.Reason
				result.Conversation = conversation
				return result
			}
		}

		response := opts.InvokeModel(conversation)
		result.Iterations++

		if opts.OutputGuardrail != nil {
			decision := opts.OutputGuardrail(response)
			if !decision.Allowed {
				result.Error = agentGuardrailError
				result.ErrorReason = decision.Reason
				result.Conversation = conversation
				return result
			}
		}

		if len(response.ToolCalls) > 0 {
			conversation = append(conversation, agentAssistantToolCallsMessage(response))
			result.ToolRounds++
			cancelled := false

			for idx, call := range response.ToolCalls {
				emit("tool_call_start", map[string]interface{}{"name": call.Name, "arguments": call.Arguments})

				if opts.ToolGuardrail != nil {
					decision := opts.ToolGuardrail(call.Name, agentParseArgs(call.Arguments))
					if !decision.Allowed {
						result.DeniedTools = append(result.DeniedTools, call.Name)
						denial := "Tool denied by guardrail: " + interfaceToString(decision.Reason)
						conversation = append(conversation, agentToolMessage(call.Id, denial))
						continue
					}
				}

				if !registered(call.Name) {
					result.Error = "Tool not registered: " + call.Name
					result.ErrorType = "ValueError"
					result.Conversation = conversation
					return result
				}

				output := opts.DispatchTool(call)
				result.ToolsExecuted++
				result.ToolExecutionOrder = append(result.ToolExecutionOrder, call.Name)
				emit("tool_result", map[string]interface{}{"name": call.Name, "result": output})
				conversation = append(conversation, agentToolMessage(call.Id, output))

				if opts.CancelAt == agentAfterToolKey(idx) {
					emit("cancelled", map[string]interface{}{"reason": "Cancellation requested after tool execution"})
					result.Error = agentCancelledError
					cancelled = true
					break
				}
			}

			if cancelled {
				result.Conversation = conversation
				return result
			}

			emit("messages_updated", map[string]interface{}{"message_count": len(conversation) + 1})

			if result.Iterations > maxIterations {
				result.Error = "Agent loop exceeded " + itoa(maxIterations) + " iterations"
				result.Conversation = conversation
				return result
			}

			continue
		}

		result.Result = response.Content
		conversation = append(conversation, map[string]interface{}{"role": "assistant", "content": response.Content})
		emit("done", map[string]interface{}{"response": response.Content})
		result.Conversation = conversation
		return result
	}
}

File: runtime/csharp/Prompty.Core/AgentLoopEngine.cs
// Copyright (c) Microsoft. All rights reserved.

using System.Text.Json;
using System.Text.Json.Nodes;

namespace Prompty.Core;

/// <summary>
/// A single tool invocation requested by the model.
/// </summary>
public sealed class AgentToolCall(string id, string name, string arguments)
{
    /// <summary>The provider-assigned tool-call id.</summary>
    public string Id { get; } = id;

    /// <summary>The tool name.</summary>
    public string Name { get; } = name;

    /// <summary>The raw JSON argument string exactly as the model emitted it.</summary>
    public string Arguments { get; } = arguments;
}

/// <summary>
/// A normalized single-turn model response. <see cref="RawToolCalls"/> carries the
/// provider's exact tool-call array so the assistant message's
/// <c>metadata.tool_calls</c> round-trips byte-for-byte; when null the engine
/// reconstructs it from <see cref="ToolCalls"/>.
/// </summary>
public sealed class AgentModelResponse
{
    /// <summary>The assistant content (final answer) when there are no tool calls.</summary>
    public string? Content { get; init; }

    /// <summary>The tool calls requested by the model, if any.</summary>
    public IReadOnlyList<AgentToolCall> ToolCalls { get; init; } = [];

    /// <summary>The provider's exact raw tool-call array, preserved verbatim.</summary>
    public JsonArray? RawToolCalls { get; init; }
}

/// <summary>Outcome of a guardrail check.</summary>
public sealed class AgentGuardrailDecision(bool allowed, string? reason = null)
{
    /// <summary>Whether the guarded action is permitted.</summary>
    public bool Allowed { get; } = allowed;

    /// <summary>The denial reason, when not allowed.</summary>
    public string? Reason { get; } = reason;
}

/// <summary>A steering message scheduled for injection before a given iteration.</summary>
public sealed class AgentSteeringMessage(int injectBeforeIteration, string role, string text)
{
    /// <summary>The 1-based iteration this message is injected before.</summary>
    public int InjectBeforeIteration { get; } = injectBeforeIteration;

    /// <summary>The message role.</summary>
    public string Role { get; } = role;

    /// <summary>The message text.</summary>
    public string Text { get; } = text;
}

/// <summary>The observable result of an agent-loop run.</summary>
public sealed class AgentLoopResult
{
    /// <summary>The final assistant answer, when the loop completed normally.</summary>
    public string? Result { get; set; }

    /// <summary>Number of LLM calls (not tool rounds).</summary>
    public int Iterations { get; set; }

    /// <summary>The accumulated conversation.</summary>
    public List<JsonObject> Conversation { get; set; } = [];

    /// <summary>The ordered event stream (<c>{type, data}</c>).</summary>
    public List<JsonObject> Events { get; } = [];

    /// <summary>Number of tool rounds executed.</summary>
    public int ToolRounds { get; set; }

    /// <summary>Number of individual tools executed.</summary>
    public int ToolsExecuted { get; set; }

    /// <summary>Ordered names of executed tools.</summary>
    public List<string> ToolExecutionOrder { get; } = [];

    /// <summary>Names of tools denied by a guardrail.</summary>
    public List<string> DeniedTools { get; } = [];

    /// <summary>The compacted conversation, when trimming occurred.</summary>
    public List<JsonObject>? TrimmedMessages { get; set; }

    /// <summary>The error marker (class name), when the loop failed.</summary>
    public string? Error { get; set; }

    /// <summary>The error type, for tool-registration errors.</summary>
    public string? ErrorType { get; set; }

    /// <summary>The guardrail denial reason, when applicable.</summary>
    public string? ErrorReason { get; set; }

    /// <summary>Conversation length plus the conformance <c>+1</c> when tools ran.</summary>
    public int TotalMessages => Conversation.Count + (ToolRounds > 0 ? 1 : 0);
}

/// <summary>
/// Provider-agnostic agent loop — the canonical <c>TurnConformance.run</c> engine.
///
/// Owns the observable agent-loop contract asserted by the cross-runtime
/// <c>@vector</c> suite (<c>schema/model/conformance/vectors/agent.tsp</c>, stage
/// <c>agent</c>). The loop is driven by two abstract callbacks —
/// <c>invokeModel(conversation)</c> and <c>dispatchTool(call)</c> — so the same
/// engine backs every provider; providers supply only wire translation. Errors are
/// returned as fields on <see cref="AgentLoopResult"/> (not thrown) so the
/// accumulated conversation and events remain observable on the failure path.
/// </summary>
public static class AgentLoopEngine
{
    /// <summary>Default maximum number of LLM iterations.</summary>
    public const int DefaultMaxIterations = 10;

    /// <summary>Canonical compaction-summary prefix.</summary>
    public const string SummaryPrefix = "[Summary of earlier conversation] ";

    // Canonical error markers mirroring the exception class names the vectors assert.
    private const string CancelledError = "CancelledError";
    private const string GuardrailErrorMarker = "GuardrailError";

    /// <summary>
    /// Run the canonical agent loop and return its observable result.
    /// </summary>
    /// <param name="messages">The starting conversation.</param>
    /// <param name="invokeModel">One LLM call, returning a normalized response.</param>
    /// <param name="dispatchTool">One tool execution, returning its string result.</param>
    /// <param name="isToolRegistered">Whether a tool name is registered (default: all).</param>
    /// <param name="maxIterations">Maximum LLM iterations.</param>
    /// <param name="inputGuardrail">Runs before each LLM call.</param>
    /// <param name="outputGuardrail">Runs on each model response.</param>
    /// <param name="toolGuardrail">Runs before each tool execution.</param>
    /// <param name="steering">Steering messages to inject before specific iterations.</param>
    /// <param name="cancelAt">Scripted cancellation position.</param>
    /// <param name="contextBudget">Character budget triggering compaction.</param>
    /// <param name="summarize">Summarizer for compaction.</param>
    public static AgentLoopResult Run(
        IEnumerable<JsonObject> messages,
        Func<List<JsonObject>, AgentModelResponse> invokeModel,
        Func<AgentToolCall, string> dispatchTool,
        Func<string, bool>? isToolRegistered = null,
        int maxIterations = DefaultMaxIterations,
        Func<List<JsonObject>, AgentGuardrailDecision>? inputGuardrail = null,
        Func<AgentModelResponse, AgentGuardrailDecision>? outputGuardrail = null,
        Func<string, JsonObject, AgentGuardrailDecision>? toolGuardrail = null,
        IReadOnlyList<AgentSteeringMessage>? steering = null,
        string? cancelAt = null,
        int? contextBudget = null,
        Func<List<JsonObject>, string>? summarize = null)
    {
        var result = new AgentLoopResult();
        var conversation = messages.Select(m => (JsonObject)m.DeepClone()).ToList();

        void Emit(string eventType, JsonObject data) =>
            result.Events.Add(new JsonObject { ["type"] = eventType, ["data"] = data });

        Emit("status", new JsonObject { ["message"] = "Starting agent loop" });

        var trimmed = MaybeTrim(conversation, contextBudget, summarize);
        if (trimmed is not null)
        {
            conversation = trimmed;
            result.TrimmedMessages = trimmed.Select(m => (JsonObject)m.DeepClone()).ToList();
        }

        var steeringPending = new List<AgentSteeringMessage>(steering ?? []);
        var registered = isToolRegistered ?? (_ => true);

        while (true)
        {
            var iterationNumber = result.Iterations + 1;

            if (cancelAt == "before_iteration" && iterationNumber == 1)
            {
                Emit("cancelled", new JsonObject { ["reason"] = "Cancellation requested before first iteration" });
                result.Error = CancelledError;
                result.Conversation = conversation;
                return result;
            }

            if (cancelAt == $"before_iteration_{iterationNumber}")
            {
                Emit("cancelled", new JsonObject { ["reason"] = $"Cancellation requested before iteration {iterationNumber}" });
                result.Error = CancelledError;
                result.Conversation = conversation;
                return result;
            }

            var toInject = steeringPending.Where(s => s.InjectBeforeIteration == iterationNumber).ToList();
            if (toInject.Count > 0)
            {
                foreach (var s in toInject)
                    steeringPending.Remove(s);
                Emit("status", new JsonObject { ["message"] = "Injecting steering message" });
                foreach (var s in toInject)
                    conversation.Add(new JsonObject { ["role"] = s.Role, ["content"] = s.Text });
                Emit("messages_updated", new JsonObject { ["message_count"] = conversation.Count + 1 });
            }

            if (inputGuardrail is not null)
            {
                var decision = inputGuardrail(conversation);
                if (!decision.Allowed)
                {
                    result.Error = GuardrailErrorMarker;
                    result.ErrorReason = decision.Reason;
                    result.Conversation = conversation;
                    return result;
                }
            }

            var response = invokeModel(conversation);
            result.Iterations += 1;

            if (outputGuardrail is not null)
            {
                var decision = outputGuardrail(response);
                if (!decision.Allowed)
                {
                    result.Error = GuardrailErrorMarker;
                    result.ErrorReason = decision.Reason;
                    result.Conversation = conversation;
                    return result;
                }
            }

            if (response.ToolCalls.Count > 0)
            {
                conversation.Add(AssistantToolCallsMessage(response));
                result.ToolRounds += 1;
                var cancelled = false;

                for (var idx = 0; idx < response.ToolCalls.Count; idx++)
                {
                    var call = response.ToolCalls[idx];
                    Emit("tool_call_start", new JsonObject { ["name"] = call.Name, ["arguments"] = call.Arguments });

                    if (toolGuardrail is not null)
                    {
                        var decision = toolGuardrail(call.Name, ParseArgs(call.Arguments));
                        if (!decision.Allowed)
                        {
                            result.DeniedTools.Add(call.Name);
                            var denial = $"Tool denied by guardrail: {decision.Reason}";
                            conversation.Add(ToolMessage(call.Id, denial));
                            continue;
                        }
                    }

                    if (!registered(call.Name))
                    {
                        result.Error = $"Tool not registered: {call.Name}";
                        result.ErrorType = "ValueError";
                        result.Conversation = conversation;
                        return result;
                    }

                    var output = dispatchTool(call);
                    result.ToolsExecuted += 1;
                    result.ToolExecutionOrder.Add(call.Name);
                    Emit("tool_result", new JsonObject { ["name"] = call.Name, ["result"] = output });
                    conversation.Add(ToolMessage(call.Id, output));

                    if (cancelAt == $"after_tool_{idx}")
                    {
                        Emit("cancelled", new JsonObject { ["reason"] = "Cancellation requested after tool execution" });
                        result.Error = CancelledError;
                        cancelled = true;
                        break;
                    }
                }

                if (cancelled)
                {
                    result.Conversation = conversation;
                    return result;
                }

                Emit("messages_updated", new JsonObject { ["message_count"] = conversation.Count + 1 });

                if (result.Iterations > maxIterations)
                {
                    result.Error = $"Agent loop exceeded {maxIterations} iterations";
                    result.Conversation = conversation;
                    return result;
                }

                continue;
            }

            result.Result = response.Content;
            conversation.Add(new JsonObject { ["role"] = "assistant", ["content"] = response.Content });
            Emit("done", new JsonObject { ["response"] = response.Content });
            result.Conversation = conversation;
            return result;
        }
    }

    private static JsonObject AssistantToolCallsMessage(AgentModelResponse response)
    {
        JsonArray toolCalls;
        if (response.RawToolCalls is not null)
        {
            toolCalls = (JsonArray)response.RawToolCalls.DeepClone();
        }
        else
        {
            toolCalls = [];
            foreach (var tc in response.ToolCalls)
            {
                toolCalls.Add(new JsonObject
                {
                    ["id"] = tc.Id,
                    ["type"] = "function",
                    ["function"] = new JsonObject { ["name"] = tc.Name, ["arguments"] = tc.Arguments },
                });
            }
        }

        return new JsonObject
        {
            ["role"] = "assistant",
            ["content"] = "",
            ["metadata"] = new JsonObject { ["tool_calls"] = toolCalls },
        };
    }

    private static JsonObject ToolMessage(string callId, string content) => new()
    {
        ["role"] = "tool",
        ["content"] = content,
        ["metadata"] = new JsonObject { ["tool_call_id"] = callId },
    };

    private static int CharCount(IEnumerable<JsonObject> messages)
    {
        var total = 0;
        foreach (var m in messages)
        {
            if (m["content"] is JsonValue value && value.TryGetValue<string>(out var content))
                total += content.Length;
        }
        return total;
    }

    private static JsonObject ParseArgs(string arguments)
    {
        if (string.IsNullOrEmpty(arguments))
            return [];
        try
        {
            return JsonNode.Parse(arguments) as JsonObject ?? [];
        }
        catch (JsonException)
        {
            return [];
        }
    }

    private static string DefaultSummary(IEnumerable<JsonObject> droppedUsers)
    {
        var topics = new List<string>();
        foreach (var m in droppedUsers)
        {
            if (m["content"] is JsonValue value && value.TryGetValue<string>(out var content))
            {
                var trimmed = content.Trim();
                if (trimmed.Length > 0)
                    topics.Add(trimmed);
            }
        }
        return SummaryPrefix + "User asked about " + string.Join("; ", topics);
    }

    private static List<JsonObject>? MaybeTrim(
        List<JsonObject> conversation,
        int? contextBudget,
        Func<List<JsonObject>, string>? summarize)
    {
        if (contextBudget is null || CharCount(conversation) <= contextBudget.Value)
            return null;

        var systems = conversation
            .Where(m => (m["role"] as JsonValue)?.GetValue<string>() == "system")
            .Select(m => (JsonObject)m.DeepClone())
            .ToList();
        var users = conversation
            .Where(m => (m["role"] as JsonValue)?.GetValue<string>() == "user")
            .ToList();
        var droppedUsers = users.Count > 0 ? users.Take(users.Count - 1).ToList() : [];
        var lastUser = users.Count > 0 ? users[^1] : null;

        var summaryText = summarize is not null ? summarize(droppedUsers) : DefaultSummary(droppedUsers);
        var summaryMessage = new JsonObject { ["role"] = "system", ["content"] = summaryText };

        var trimmed = new List<JsonObject>(systems) { summaryMessage };
        if (lastUser is not null)
            trimmed.Add(new JsonObject { ["role"] = "user", ["content"] = lastUser["content"]?.DeepClone() });
        return trimmed;
    }
}
